api/layer/tagging.go

```diff
@@ -0,0 +1,169 @@
+	var (
+		err  error
+		tags map[string]string
+	)
+	tags = n.systemCache.GetTagging(objectTaggingCacheKey(p))
+	if tags != nil {
+		return tags, nil
+	}
+
+	version, err := n.getTaggedObjectVersion(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	tags, err = n.treeService.GetObjectTagging(ctx, p.CnrID, version)
+	if err != nil {
+		if errorsStd.Is(err, ErrNodeNotFound) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchKey)
+		}
+		return nil, err
+	}
+
+	if err = n.systemCache.PutTagging(objectTaggingCacheKey(p), tags); err != nil {
+		n.log.Error("couldn't cache system object", zap.Error(err))
+	}
+
+	return tags, nil
+}
+
+func (n *layer) PutObjectTagging(ctx context.Context, p *data.ObjectTaggingInfo, tagSet map[string]string) error {
+	version, err := n.getTaggedObjectVersion(ctx, p)
+	if err != nil {
+		return err
+	}
+
+	err = n.treeService.PutObjectTagging(ctx, p.CnrID, version, tagSet)
+	if err != nil {
+		if errorsStd.Is(err, ErrNodeNotFound) {
+			return errors.GetAPIError(errors.ErrNoSuchKey)
+		}
+		return err
+	}
+
+	if err = n.systemCache.PutTagging(objectTaggingCacheKey(p), tagSet); err != nil {
+		n.log.Error("couldn't cache system object", zap.Error(err))
+	}
+
+	return nil
+}
+
+func (n *layer) DeleteObjectTagging(ctx context.Context, p *data.ObjectTaggingInfo) error {
+	version, err := n.getTaggedObjectVersion(ctx, p)
+	if err != nil {
+		return err
+	}
+
+	err = n.treeService.DeleteObjectTagging(ctx, p.CnrID, version)
+	if err != nil {
+		if errorsStd.Is(err, ErrNodeNotFound) {
+			return errors.GetAPIError(errors.ErrNoSuchKey)
+		}
+		return err
+	}
+
+	n.systemCache.Delete(objectTaggingCacheKey(p))
+
+	return nil
+}
+
+func (n *layer) GetBucketTagging(ctx context.Context, cnrID *cid.ID) (map[string]string, error) {
+	var (
+		err  error
+		tags map[string]string
+	)
+
+	tags = n.systemCache.GetTagging(bucketTaggingCacheKey(cnrID))
+	if tags != nil {
+		return tags, nil
+	}
+
+	if tags, err = n.treeService.GetBucketTagging(ctx, cnrID); err != nil && !errorsStd.Is(err, ErrNodeNotFound) {
+		return nil, err
+	}
+
+	if err := n.systemCache.PutTagging(bucketTaggingCacheKey(cnrID), tags); err != nil {
+		n.log.Error("couldn't cache system object", zap.Error(err))
+	}
+
+	return tags, nil
+}
+
+func (n *layer) PutBucketTagging(ctx context.Context, cnrID *cid.ID, tagSet map[string]string) error {
+	if err := n.treeService.PutBucketTagging(ctx, cnrID, tagSet); err != nil {
+		return err
+	}
+	if err := n.systemCache.PutTagging(bucketTaggingCacheKey(cnrID), tagSet); err != nil {
+		n.log.Error("couldn't cache system object", zap.Error(err))
+	}
+
+	return nil
+}
+
+func (n *layer) DeleteBucketTagging(ctx context.Context, cnrID *cid.ID) error {
+	n.systemCache.Delete(bucketTaggingCacheKey(cnrID))
+
+	return n.treeService.DeleteBucketTagging(ctx, cnrID)
+}
+
+func objectTaggingCacheKey(p *data.ObjectTaggingInfo) string {
+	return ".tagset." + p.CnrID.String() + "." + p.ObjName + "." + p.VersionID
```

Comment:
Please use `CnrID.EncodeToString()` instead of `cnrID.String()` in this file, see https://github.com/nspcc-dev/neofs-s3-gw/issues/446